### PR TITLE
[model, ops] feat: add Qwen3 sequence classification model and loss for embedding classification.

### PR DESCRIPTION
**Pull request by yiwzhao**
feat: Add a data collator to support embedding classification.

### CODE REVIEW

File: veomni/data/data_collator.py [hunk 1]
```diff
@@ -325,3 +325,120 @@ def __call__(self, batch: Sequence[Dict[str, "torch.Tensor"]]) -> Dict[str, "tor
             add_flash_attention_kwargs_from_position_ids(batch)
 
         return batch
+
+
+@dataclass
+class ClassificationDataCollatorWithPositionIDs(DataCollator):
```
Comment (piyifan123) [inline veomni/data/data_collator.py]: split this to another MR

File: veomni/models/loader.py [hunk 2]
```diff
@@ -114,6 +112,8 @@ def get_model_arch_from_config(model_config):
 
     arch_name = get_model_arch_from_config(model_config)
     model_type = model_config.model_type
+    if not force_use_huggingface:
```
Comment (piyifan123) [inline veomni/models/loader.py]: rebase this?

File: veomni/ops/loss.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    hidden_states: torch.Tensor,
+    weight: torch.Tensor,
+    labels: torch.Tensor,
+    num_items_in_batch: Optional[int] = None,
+    ignore_index: int = -100,
+    shift_labels: Optional[torch.Tensor] = None,
+    **kwargs,
+) -> torch.Tensor:
+    # We don't use shift_labels
+    assert shift_labels is None
```
Comment (piyifan123) [inline veomni/ops/loss.py]: assert can be skipped. do not use in production code.

File: veomni/ops/loss.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    loss = None
+    logits = None
+
+    if labels is None:
```
Comment (piyifan123) [inline veomni/ops/loss.py]: throw exception if label is None.

File: veomni/ops/loss.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Reduce loss when using sp
+    if sp_enabled:
+        num_valid_tokens = (labels != IGNORE_INDEX).sum()
+        loss = reduce_sequence_parallel_loss(loss, num_valid_tokens)
+
+    return loss, logits
+
+
+def seqcls_token_loss_function(
```
Comment (piyifan123) [inline veomni/ops/loss.py]: this file is no longer there. can we follow the new way defined in https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/ops/fused_cross_entropy/__init__.py

File: veomni/ops/loss.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        return loss, logits
+
+    sp_enabled = get_parallel_state().sp_enabled
+
+    # Flatten the labels and hidden_states
+    labels = labels.view(-1)
+    hidden_states = hidden_states.view(-1, hidden_states.size(-1))
+
+    # Calculate loss
+    if fused_linear_cross_entropy is not None:  # use kernels
```
Comment (piyifan123) [inline veomni/ops/loss.py]: follow the impl in https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/ops/fused_cross_entropy/__init__.py#L78-L86

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 7]
```diff
@@ -737,15 +725,74 @@ def forward(
...
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            **kwargs,
+        )
+
+        hidden_states = transformer_outputs.last_hidden_state
+        logits = self.score(hidden_states)
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py]: remove this. no longer needed. we can just use the one from the loss_function.

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 8]
```diff
@@ -737,15 +725,74 @@ def forward(
...
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Cache] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> SequenceClassifierOutputWithPast:
+        transformer: Qwen3Model = getattr(self, self.base_model_prefix)
+        transformer_outputs: BaseModelOutputWithPast = transformer(
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py]: self.model(...)?

File: veomni/ops/loss.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    return loss, logits
+
+
+def seqcls_token_loss_function(
+    hidden_states: torch.Tensor,
+    weight: torch.Tensor,
+    labels: torch.Tensor,
+    num_items_in_batch: Optional[int] = None,
+    ignore_index: int = -100,
+    shift_labels: Optional[torch.Tensor] = None,
```
Comment (piyifan123) [inline veomni/ops/loss.py]: remove shift_labels. kwargs

File: tests/data/test_seqcls_loss.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    H = 8
+    C = 5
+    B, L = 2, 3
+    hidden_states = torch.randn(B, L, H)
+    weight = torch.randn(C, H)
+    labels = torch.tensor([[1, -100, 2], [3, 4, -100]], dtype=torch.long)
+
+    loss, logits = m.seqcls_token_loss_function(hidden_states, weight, labels=labels, ignore_index=-100)
+
+    assert loss is not None
```
Comment (piyifan123) [inline tests/data/test_seqcls_loss.py]: https://docs.google.com/document/d/13k8AsgYdF-2TJx9QIHHvPiuV3JMbsrMLkWmmjdYBLVg/edit?tab=t.0

File: veomni/ops/loss.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    **kwargs,
+) -> torch.Tensor:
```
Comment (piyifan123) [inline veomni/ops/loss.py]: add docstring.

File: veomni/models/loader.py (hunk 2 above)
Comment (yiwzhao) [reply veomni/models/loader.py]: Already rebased.

File: veomni/ops/loss.py (hunk 3 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: resolved.

File: veomni/ops/loss.py (hunk 4 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: It has been implemented that a ValueError will be raised if the label is none now.

File: veomni/ops/loss.py (hunk 5 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: Yes, we can. Implemented using the latest way.

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 7 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py]: removed.

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 8 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py]: It has been revised to a simpler version as suggested.

File: veomni/ops/loss.py (hunk 9 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: removed.

File: veomni/data/data_collator.py (hunk 1 above)
Comment (yiwzhao) [reply veomni/data/data_collator.py]: Moved to https://github.com/ByteDance-Seed/VeOmni/pull/376

File: veomni/ops/loss.py (hunk 6 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: Done.

File: veomni/ops/loss.py (hunk 11 above)
Comment (yiwzhao) [reply veomni/ops/loss.py]: added.

File: tests/data/test_seqcls_loss.py (hunk 10 above)
Comment (yiwzhao) [reply tests/data/test_seqcls_loss.py]: The unit tests for the loss function have been adjusted according to this document

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 12]
```diff
@@ -737,6 +738,94 @@ def forward(
...
+        if labels is not None:
+            loss, logits = self.loss_function(
+                logits=logits,
+                labels=labels,
+                vocab_size=self.num_labels,
+                hidden_states=hidden_states,
+                weights=self.score.weight,
+                **kwargs,
+            )
+        else:
```
Comment (Coach257) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:819]: what if inference task

Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:819]: Added logic to calculate logits when no labels are provided, for compatibility with inference tasks.

File: tests/ops/test_seqcls_loss.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
+import math
```
Comment (Luosuu) [inline tests/ops/test_seqcls_loss.py]: tbh I dont understand what this test does..

File: veomni/ops/fused_cross_entropy/__init__.py [hunk 14]
```diff
@@ -104,6 +104,98 @@ def ForCausalLMLoss(
...
+    target = labels
+
+    # Flatten the tokens
+    target = target.view(-1).to(device)
+    if hidden_states is not None:
+        hidden_states = hidden_states.view(-1, hidden_states.size(-1))
+    if logits is not None:
+        logits = logits.view(-1, num_labels)
+    # Enable model parallelism
+    target = target.to(device)
```
Comment (Luosuu) [inline veomni/ops/fused_cross_entropy/__init__.py:175]: could you please explain why this is related to model parallelism?

File: tests/ops/test_seqcls_loss.py (hunk 13 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: It verifies that the seq-classification loss uses the right function, handles masking and SP correctly, and produces the exact cross-entropy value expected from a manual calculation. Currently, this only involves manually constructing test cases and verifying whether the loss value is calculated correctly.  Perhaps in the future, we can add a real test within the trainer.

File: veomni/models/loader.py [hunk 15]
```diff
@@ -114,6 +115,9 @@ def get_model_arch_from_config(model_config):
 
     arch_name = get_model_arch_from_config(model_config)
     model_type = model_config.model_type
+    modeling_backend = get_env("MODELING_BACKEND")
+    if not modeling_backend == "hf":
```
Comment (piyifan123) [inline veomni/models/loader.py:119]: @Coach257 can you take a look? i think it's a cleaner and equivalent impl as before, i.e., load_class needs to be computed only when modeling_backend is hf. otherwise we don't need to calculate load_class but return MODELING_REGISTRY[model_type](arch_name) directly.

File: veomni/ops/fused_cross_entropy/__init__.py [hunk 16]
```diff
@@ -104,6 +104,100 @@ def ForCausalLMLoss(
...
+            Scalar classification loss.
+        logits (`torch.Tensor`):
+            Flattened logits.
+    """
+
+    # pop fused loss kwargs
+    hidden_states = kwargs.pop("hidden_states", None)
+    weights = kwargs.pop("weights", None)
+
+    assert hidden_states is not None or logits is not None, "hidden_states or logits must be provided."
```
Comment (piyifan123) [inline veomni/ops/fused_cross_entropy/__init__.py]: https://google.github.io/styleguide/pyguide.html#24-exceptions

File: tests/ops/test_seqcls_loss.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+def test_seqcls_loss_prefers_cross_entropy_when_hidden_states_and_weights_present(monkeypatch):
+    """
+    logits provided
+    hidden_states+weights present
+    sp_enabled = False
+    """
+    device = torch.device("cuda")
+    monkeypatch.setattr(m, "get_parallel_state", lambda: _FakePS(sp_enabled=False))
+    ignore = -100
```
Comment (piyifan123) [inline tests/ops/test_seqcls_loss.py]: import veomni constant IGNORE_INDEX

File: tests/ops/test_seqcls_loss.py [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+
+    loss, out_logits = m.ForSequenceClassificationLoss(
+        logits=logits,
+        labels=labels,
+        num_labels=3,
+        ignore_index=ignore,
+        hidden_states=hidden_states,
+        weights=weights,
+    )
+    expected = math.log(float(3))
```
Comment (piyifan123) [inline tests/ops/test_seqcls_loss.py]: can we just write down the one line torch command to do the matmul + softmax + cross entropy

File: tests/ops/test_seqcls_loss.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+    hidden_states+weights present
+    sp_enabled = False
+    """
+    device = torch.device("cuda")
+    monkeypatch.setattr(m, "get_parallel_state", lambda: _FakePS(sp_enabled=False))
+    ignore = -100
+    logits = torch.zeros((1, 2, 3), device=device)
+    labels = torch.tensor([[ignore, 1]], device=device)
+    hidden_states = torch.zeros((1, 2, 5), device=device)
+    weights = torch.zeros((3, 5), device=device)
```
Comment (piyifan123) [inline tests/ops/test_seqcls_loss.py]: can we make it a matrix.

File: tests/ops/test_seqcls_loss.py (hunk 13 above)
Comment (piyifan123) [inline tests/ops/test_seqcls_loss.py]: check if we need to add tests/ops folder to https://github.com/ByteDance-Seed/VeOmni/blob/main/.github/workflows/gpu_unit_tests.yml

File: veomni/models/loader.py (hunk 15 above)
Comment (Coach257) [reply veomni/models/loader.py:119]: I agree the change here.

File: tests/ops/test_seqcls_loss.py (hunk 13 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: It verifies that the seq-classification loss uses the right function, handles masking and SP correctly, and produces the exact cross-entropy value expected from a manual calculation. Currently, this only involves manually constructing test cases and verifying whether the loss value is calculated correctly. Perhaps in the future, we can add a real test within the trainer.

File: tests/ops/test_seqcls_loss.py (hunk 17 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: replaced

File: veomni/ops/fused_cross_entropy/__init__.py (hunk 16 above)
Comment (yiwzhao) [reply veomni/ops/fused_cross_entropy/__init__.py]: Replaced asserts with explicit ValueError.

File: tests/ops/test_seqcls_loss.py (hunk 13 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: Yes, we need it. I added the tests/ops directory.

File: tests/ops/test_seqcls_loss.py (hunk 19 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: Yes, sure. Updated.

File: tests/ops/test_seqcls_loss.py (hunk 18 above)
Comment (yiwzhao) [reply tests/ops/test_seqcls_loss.py]: Updated.
